---
title: "Infernal tRNA report summary"
author: "Teitur Ahlgren Kalman"
date: "`r Sys.Date()`"
output:
 html_document:
   toc: true
   number_sections: false
   code_folding: "hide"
---
# Libraries 

````{r }
suppressPackageStartupMessages({
  library(systemPipeR)
  library(ggplot2)
  library(stringr)
  library(plyr)
})
````

# Description
Looking at the where tDNA sequences are found in the new P.abies assembly.
To produce this report I intersected 10kb windows of the genome with locations 
of the tDNA sequences found with Infernal.
```{r set up, echo=FALSE}
knitr::opts_knit$set(root.dir="/mnt/picea/home/tkalman")
```
# Data paths

````{r }
Pabies.10kb_window.tRNA_intersect.path  <- "/mnt/picea/home/tkalman/tRNA-rRNA/tRNA_seq/data/intersect/Pabies.10000bp-windows.bed.bedtools-intersect.tRNA-id90.cmsearch.sorted.bed.tsv"
Pabies.unplaced.tRNA_intersect.path <- "/mnt/picea/home/tkalman/tRNA-rRNA/tRNA_seq/data/intersect/unplaced_contiglengths.bed.bedtools-intersect.tRNA-id90.cmsearch.sorted.bed.tsv"
````

# Any overrepresentation of tRNA on any of the chromosomes?
Plot the number of matches per chromosome ID

````{r }
tRNA.chr_count.df <- data.frame(do.call(rbind, str_split(system(paste("awk '{if ($10 > 0) print $1}'", Pabies.10kb_window.tRNA_intersect.path, "| sort | uniq -c | awk '{print $1\"\t\"$2}'"),
                                                                intern = TRUE), 
                                                         pattern = "\t")))

ggplot(tRNA.chr_count.df, aes(x = X2, y = as.numeric(X1))) +
  geom_col() +
  ggtitle("tRNA match frequency per chromosome") +
  xlab("Chr ID") +
  ylab("Match count") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
````

# Plot sliding window
Define plotting function

````{r }
every_nth = function(n) {
  return(function(x) {x[c(TRUE, rep(FALSE, n - 1))]})
}

plot.sliding_window_intersect <- function(intersect.path, contig.ID) {
  tmp.df <- data.frame(do.call(rbind, strsplit(system(paste("grep -w", contig.ID, intersect.path, "| awk '{print $3\"\t\"$10}'"), intern = TRUE), "\t")))
  tmp.df <- ddply(tmp.df, .(X1), summarise, X2=sum(as.numeric(X2)))
  ggplot(tmp.df, aes(x = as.numeric(X1), y = as.numeric(X2), group = 1)) +
    geom_point() +
    ggtitle(paste("tRNA feature overlap per 10kb window in:", contig.ID)) +
    ylab("length of overlap (bp)") +
    xlab("chromosome coordinate (bp)") +
    theme_minimal() +
    theme(axis.text.x = element_text(size = 7, angle = 45, hjust = 1))
}
````

Plot

````{r }
plot.sliding_window_intersect(Pabies.10kb_window.tRNA_intersect.path,
                              "PA_chr01")

plot.sliding_window_intersect(Pabies.10kb_window.tRNA_intersect.path,
                              "PA_chr02")

plot.sliding_window_intersect(Pabies.10kb_window.tRNA_intersect.path,
                              "PA_chr03")

plot.sliding_window_intersect(Pabies.10kb_window.tRNA_intersect.path,
                              "PA_chr04")

plot.sliding_window_intersect(Pabies.10kb_window.tRNA_intersect.path,
                              "PA_chr05")

plot.sliding_window_intersect(Pabies.10kb_window.tRNA_intersect.path,
                              "PA_chr06")

plot.sliding_window_intersect(Pabies.10kb_window.tRNA_intersect.path,
                              "PA_chr07")

plot.sliding_window_intersect(Pabies.10kb_window.tRNA_intersect.path,
                              "PA_chr08")

plot.sliding_window_intersect(Pabies.10kb_window.tRNA_intersect.path,
                              "PA_chr09")

plot.sliding_window_intersect(Pabies.10kb_window.tRNA_intersect.path,
                              "PA_chr10")

plot.sliding_window_intersect(Pabies.10kb_window.tRNA_intersect.path,
                              "PA_chr11")

plot.sliding_window_intersect(Pabies.10kb_window.tRNA_intersect.path,
                              "PA_chr12")
````

# Analysis of tRNA in unplaced
There are 882 contigs of unplaced sequences, looking at all of them at the same time
since these are of uneven length I normalize with contiglengths
Define function for plotting

````{r }
plot.sliding_window_intersect <- function(intersect.path) {
  tmp.df <- data.frame(do.call(rbind, strsplit(system(paste("cat", intersect.path, "| awk '{print $1\"\t\"$10}'"), intern = TRUE), "\t")))
  tmp.df <- ddply(tmp.df, .(X1), summarise, X2=sum(as.numeric(X2)))
  ggplot(tmp.df, aes(x = X1, y = as.numeric(X2), group = 1)) +
    geom_point() +
    scale_x_discrete(breaks = every_nth(n = 20)) +
    ggtitle(paste("tRNA feature overlap in unplaced")) +
    ylab("length of overlap (bp)") +
    xlab("chromosome coordinate (bp)") +
    theme_minimal() +
    theme(axis.text.x = element_text(size = 7, angle = 90, hjust = 1))
}

plot.sliding_window_intersect(Pabies.unplaced.tRNA_intersect.path)
````

Possibly use tRNA matches to map unplaced contigs in assembly?
